File: modules/03-deep-learning/01-neural-networks/01-mlp-forward-backward/starter/starter.py
# Dependencies: torch>=2.1, numpy>=1.24, matplotlib>=3.7, scikit-learn>=1.3
# Node: 3.1.1 — Multi-Layer Perceptrons — Forward & Backward Pass
# Run: python starter.py

import numpy as np
import matplotlib.pyplot as plt
import torch
import torch.nn as nn
from torch.utils.data import DataLoader, TensorDataset
from sklearn.datasets import make_moons
from sklearn.model_selection import train_test_split
from typing import List, Tuple


# ── Implemented Helpers ───────────────────────────────────────────────────────


def load_moons_data(
    n_samples: int = 1000,
    noise: float = 0.2,
    random_state: int = 42,
) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
    """Generate moons dataset and return train/test float32 tensors."""
    X, y = make_moons(n_samples=n_samples, noise=noise, random_state=random_state)
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=random_state
    )
    X_train_t = torch.tensor(X_train, dtype=torch.float32)
    X_test_t = torch.tensor(X_test, dtype=torch.float32)
    # Unsqueeze to shape (N, 1) — required by BCELoss
    y_train_t = torch.tensor(y_train, dtype=torch.float32).unsqueeze(1)
    y_test_t = torch.tensor(y_test, dtype=torch.float32).unsqueeze(1)
    return X_train_t, X_test_t, y_train_t, y_test_t


def plot_loss_curve(train_losses: List[float], val_losses: List[float]) -> None:
    """Save training and validation loss curves to loss_curve.png."""
    plt.figure(figsize=(8, 4))
    plt.plot(train_losses, label="Train Loss", linewidth=2)
    plt.plot(val_losses, label="Val Loss", linewidth=2, linestyle="--")
    plt.xlabel("Epoch")
    plt.ylabel("BCE Loss")
    plt.title("MLP Training Loss — 3.1.1")
    plt.legend()
    plt.tight_layout()
    plt.savefig("loss_curve.png", dpi=120)
    plt.close()
    print("Loss curve saved → loss_curve.png")


def plot_decision_boundary(
    model: nn.Module, X: torch.Tensor, y: torch.Tensor
) -> None:
    """Save 2D decision boundary contour plot to decision_boundary.png."""
    model.eval()
    x_min = X[:, 0].min().item() - 0.5
    x_max = X[:, 0].max().item() + 0.5
    y_min = X[:, 1].min().item() - 0.5
    y_max = X[:, 1].max().item() + 0.5
    xx, yy = np.meshgrid(
        np.linspace(x_min, x_max, 300),
        np.linspace(y_min, y_max, 300),
    )
    grid = torch.tensor(np.c_[xx.ravel(), yy.ravel()], dtype=torch.float32)
    with torch.no_grad():
        probs = model(grid).numpy().reshape(xx.shape)
    plt.figure(figsize=(7, 5))
    plt.contourf(xx, yy, probs, levels=50, cmap="RdBu", alpha=0.8)
    plt.colorbar(label="P(class = 1)")
    y_np = y.squeeze().numpy()
    plt.scatter(
        X[:, 0].numpy(),
        X[:, 1].numpy(),
        c=y_np,
        cmap="RdBu",
        edgecolors="k",
        s=20,
    )
    plt.title("MLP Decision Boundary — 3.1.1")
    plt.tight_layout()
    plt.savefig("decision_boundary.png", dpi=120)
    plt.close()
    print("Decision boundary saved → decision_boundary.png")


# ── Student Tasks ─────────────────────────────────────────────────────────────


def build_mlp(input_dim: int, hidden_dim: int, output_dim: int) -> nn.Module:
    """
    Task 1: Build an MLP with two hidden layers and a sigmoid output.

    Return an nn.Sequential with EXACTLY this architecture:
        Linear(input_dim  → hidden_dim) → ReLU
        Linear(hidden_dim → hidden_dim) → ReLU
        Linear(hidden_dim → output_dim) → Sigmoid

    The model accepts batches of shape (N, input_dim) and outputs (N, output_dim).
    Do NOT add BatchNorm or Dropout — keep it minimal for this exercise.
    """
    raise NotImplementedError("TODO: implement this")


def train_one_epoch(
    model: nn.Module,
    optimizer: torch.optim.Optimizer,
    criterion: nn.Module,
    X: torch.Tensor,
    y: torch.Tensor,
    batch_size: int = 32,
) -> float:
    """
    Task 2: Run one full epoch of mini-batch gradient descent.

    Use TensorDataset + DataLoader (both already imported) to iterate over
    batches of size `batch_size`. For EACH batch:
        1. optimizer.zero_grad()           — clear stale gradients
        2. predictions = model(X_batch)    — forward pass
        3. loss = criterion(predictions, y_batch)  — compute BCE loss
        4. loss.backward()                 — backprop through the graph
        5. optimizer.step()                — update weights

    Set model.train() at the start of this function.
    Return the mean loss (float) across all batches in the epoch.
    """
    raise NotImplementedError("TODO: implement this")


def evaluate_accuracy(
    model: nn.Module, X: torch.Tensor, y: torch.Tensor
) -> float:
    """
    Task 3: Compute binary classification accuracy (no gradient tracking).

    - Call model.eval() to disable dropout/batchnorm (good habit even without them).
    - Wrap the forward pass in torch.no_grad() to save memory.
    - Threshold: model output >= 0.5 → predict class 1, else class 0.
    - Return accuracy as a float in [0, 1].

    Acceptance threshold: >= 0.90 on the test set after 50 epochs.
    """
    raise NotImplementedError("TODO: implement this")


def run_training(
    model: nn.Module,
    optimizer: torch.optim.Optimizer,
    criterion: nn.Module,
    X_train: torch.Tensor,
    y_train: torch.Tensor,
    X_val: torch.Tensor,
    y_val: torch.Tensor,
    epochs: int = 50,
) -> Tuple[List[float], List[float]]:
    """
    Task 4: Run the full training loop for `epochs` epochs.

    Each epoch:
        a. Call train_one_epoch → get train loss for this epoch.
        b. Set model.eval(), compute val loss inside torch.no_grad():
               val_preds = model(X_val)
               val_loss  = criterion(val_preds, y_val).item()
        c. Print exactly:
               f"Epoch {epoch:2d}/{epochs} | Train Loss: {train_loss:.4f} | Val Loss: {val_loss:.4f}"

    Return (train_losses, val_losses) — one float per epoch, two lists of length `epochs`.
    """
    raise NotImplementedError("TODO: implement this")


# ── Main ──────────────────────────────────────────────────────────────────────


def main() -> None:
    torch.manual_seed(42)
    np.random.seed(42)

    print("=== 3.1.1 — Build basic PyTorch MLP ===\n")

    # Load synthetic moons dataset
    X_train, X_test, y_train, y_test = load_moons_data(n_samples=1000, noise=0.2)
    print(f"Train samples: {X_train.shape[0]} | Test samples: {X_test.shape[0]}\n")

    # Task 1 — build model (input_dim=2 features, hidden_dim=64, output_dim=1 probability)
    model = build_mlp(input_dim=2, hidden_dim=64, output_dim=1)
    print(f"Model architecture:\n{model}\n")

    # Optimizer and loss — Adam + BCELoss are the standard pair for this setup
    optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)
    criterion = nn.BCELoss()

    # Tasks 2, 3, 4 — full training loop
    train_losses, val_losses = run_training(
        model,
        optimizer,
        criterion,
        X_train,
        y_train,
        X_test,
        y_test,
        epochs=50,
    )

    # Task 3 — final accuracy evaluation
    accuracy = evaluate_accuracy(model, X_test, y_test)
    print(f"\nFinal Test Accuracy: {accuracy:.4f}")
    print(f"Final Train Loss:    {train_losses[-1]:.4f}")
    print(f"Final Val Loss:      {val_losses[-1]:.4f}")

    # Implemented helpers — save visualizations
    plot_loss_curve(train_losses, val_losses)
    plot_decision_boundary(model, X_test, y_test)


if __name__ == "__main__":
    main()

File: modules/03-deep-learning/01-neural-networks/01-mlp-forward-backward/starter/starter.java
// Dependencies (Maven):
// <dependency>
//   <groupId>ai.djl</groupId>
//   <artifactId>api</artifactId>
//   <version>0.26.0</version>
// </dependency>
// <dependency>
//   <groupId>ai.djl.pytorch</groupId>
//   <artifactId>pytorch-engine</artifactId>
//   <version>0.26.0</version>
// </dependency>
// <dependency>
//   <groupId>ai.djl.pytorch</groupId>
//   <artifactId>pytorch-native-auto</artifactId>
//   <version>2.1.1</version>
// </dependency>
//
// Node: 3.1.1 — Multi-Layer Perceptrons — Forward & Backward Pass
// Run (after Maven build):
//   mvn compile exec:java -Dexec.mainClass="StarterAssignment"

import ai.djl.Model;
import ai.djl.ndarray.NDArray;
import ai.djl.ndarray.NDList;
import ai.djl.ndarray.NDManager;
import ai.djl.ndarray.types.DataType;
import ai.djl.ndarray.types.Shape;
import ai.djl.nn.Activation;
import ai.djl.nn.Parameter;
import ai.djl.nn.SequentialBlock;
import ai.djl.nn.core.Linear;
import ai.djl.training.DefaultTrainingConfig;
import ai.djl.training.EasyTrain;
import ai.djl.training.Trainer;
import ai.djl.training.TrainingConfig;
import ai.djl.training.dataset.ArrayDataset;
import ai.djl.training.evaluator.BinaryAccuracy;
import ai.djl.training.listener.TrainingListener;
import ai.djl.training.loss.Loss;
import ai.djl.training.optimizer.Optimizer;
import ai.djl.training.tracker.Tracker;
import ai.djl.translate.TranslateException;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

/**
 * Starter code for Assignment 3.1.1 — Build Basic PyTorch MLP (DJL / Java).
 *
 * <p>Complete the three TODO methods in order. Each method has explicit hints showing
 * which DJL APIs to use. Run after every task to see incremental output.
 *
 * <p>Java equivalent of the Python PyTorch MLP using Deep Java Library (DJL).
 */
public class StarterAssignment {

    // ── Implemented Helper ────────────────────────────────────────────────────

    /**
     * Generate a synthetic two-moons dataset as DJL NDArrays.
     *
     * <p>Two interleaved half-circles: class 0 = upper arc, class 1 = lower arc shifted by (1,
     * 0.5). Both classes have Gaussian noise added.
     *
     * @param manager   NDManager that owns the returned arrays
     * @param nSamples  total samples (split evenly between classes)
     * @param noise     standard deviation of Gaussian noise
     * @param seed      random seed for reproducibility
     * @return float[] array {X_flat, y_flat} where X_flat has shape (nSamples, 2) flattened
     *         row-major and y_flat has shape (nSamples, 1) flattened
     */
    public static NDArray[] generateMoonsData(
            NDManager manager, int nSamples, float noise, long seed) {
        Random rng = new Random(seed);
        int half = nSamples / 2;
        float[] xData = new float[nSamples * 2];
        float[] yData = new float[nSamples];

        // Class 0 — upper half-circle (angles 0 → π)
        for (int i = 0; i < half; i++) {
            double angle = Math.PI * i / half;
            xData[i * 2]     = (float) Math.cos(angle) + (float) (rng.nextGaussian() * noise);
            xData[i * 2 + 1] = (float) Math.sin(angle) + (float) (rng.nextGaussian() * noise);
            yData[i] = 0.0f;
        }
        // Class 1 — lower half-circle (angles 0 → π), shifted right and down
        for (int i = 0; i < half; i++) {
            double angle = Math.PI * i / half;
            int idx = half + i;
            xData[idx * 2]     = (float) (1.0 - Math.cos(angle)) + (float) (rng.nextGaussian() * noise);
            xData[idx * 2 + 1] = (float) (0.5 - Math.sin(angle)) + (float) (rng.nextGaussian() * noise);
            yData[idx] = 1.0f;
        }

        NDArray X = manager.create(xData, new Shape(nSamples, 2));
        NDArray y = manager.create(yData, new Shape(nSamples, 1));
        return new NDArray[]{X, y};
    }

    /** Print a labeled metric value with consistent formatting. */
    public static void printMetric(String label, float value) {
        System.out.printf("%-30s %.4f%n", label + ":", value);
    }

    // ── Task 1 ────────────────────────────────────────────────────────────────

    /**
     * Task 1: Build an MLP using DJL's SequentialBlock.
     *
     * <p>Return a SequentialBlock with EXACTLY this architecture:
     * <pre>
     *   Linear(hiddenDim)  → ReLU
     *   Linear(hiddenDim)  → ReLU
     *   Linear(outputDim)  → Sigmoid
     * </pre>
     *
     * <p>DJL hint — SequentialBlock construction:
     * <pre>{@code
     *   SequentialBlock block = new SequentialBlock();
     *   block.add(Linear.builder().setUnits(hiddenDim).build());
     *   block.add(Activation::relu);
     *   block.add(Linear.builder().setUnits(hiddenDim).build());
     *   block.add(Activation::relu);
     *   block.add(Linear.builder().setUnits(outputDim).build());
     *   block.add(Activation::sigmoid);
     *   return block;
     * }</pre>
     *
     * <p>Note: DJL Linear layers infer inputDim automatically from the first batch shape.
     * The Trainer.initialize() call in main() triggers this inference.
     *
     * @param hiddenDim number of neurons in each hidden layer
     * @param outputDim number of output neurons (1 for binary classification)
     * @return configured SequentialBlock (weights not yet allocated)
     */
    public static SequentialBlock buildMlp(int hiddenDim, int outputDim) {
        // TODO: implement this
        throw new UnsupportedOperationException("TODO: implement this — see Javadoc above");
    }

    // ── Task 2 ────────────────────────────────────────────────────────────────

    /**
     * Task 2: Run the full training loop using DJL's EasyTrain API.
     *
     * <p>DJL wraps the mini-batch loop inside {@link EasyTrain#fit}. Your job is to
     * build the {@link ArrayDataset}, call {@code EasyTrain.fit()}, and collect the
     * per-epoch training loss from the {@link Trainer} metrics.
     *
     * <p>Steps:
     * <ol>
     *   <li>Build an ArrayDataset from X (data) and y (labels) with the given batchSize.</li>
     *   <li>Call {@code dataset.prepare()} to load it into memory.</li>
     *   <li>For each epoch 1..epochs, call {@code EasyTrain.trainDataset(trainer, dataset)}.
     *       After each epoch call {@code trainer.notifyListeners(l -> l.onEpoch(trainer))}
     *       and read the train loss via
     *       {@code trainer.getTrainingResult().getTrainLoss()}.</li>
     *   <li>Print each epoch: {@code "Epoch e/epochs | Train Loss: X.XXXX"}</li>
     * </ol>
     *
     * <p>DJL hint — ArrayDataset construction:
     * <pre>{@code
     *   ArrayDataset dataset = new ArrayDataset.Builder()
     *       .setData(X)
     *       .optLabels(y)
     *       .setSampling(batchSize, true)   // true = shuffle each epoch
     *       .build();
     * }</pre>
     *
     * @param trainer   initialized DJL Trainer bound to the model block
     * @param X         feature tensor, shape (nSamples, 2)
     * @param y         label tensor,   shape (nSamples, 1)
     * @param epochs    number of full dataset passes
     * @param batchSize mini-batch size
     * @return list of per-epoch training losses (length == epochs)
     */
    public static List<Float> trainModel(
            Trainer trainer, NDArray X, NDArray y, int epochs, int batchSize)
            throws IOException, TranslateException {
        // TODO: implement this
        throw new UnsupportedOperationException("TODO: implement this — see Javadoc above");
    }

    // ── Task 3 ────────────────────────────────────────────────────────────────

    /**
     * Task 3: Compute binary classification accuracy without gradient tracking.
     *
     * <p>Steps:
     * <ol>
     *   <li>Call {@code model.getBlock().forward(params, new NDList(X), false)} where
     *       {@code params} is {@code model.getBlock().getParameters()} and {@code false}
     *       disables training mode (no gradient computation).</li>
     *   <li>Extract the output NDArray via {@code .singletonOrThrow()}.</li>
     *   <li>Threshold at 0.5: predicted class = {@code (output >= 0.5) ? 1 : 0}.</li>
     *   <li>Compare predictions to y element-wise and compute the fraction that match.</li>
     * </ol>
     *
     * <p>DJL hint — element-wise threshold and comparison:
     * <pre>{@code
     *   NDArray predicted = output.gte(0.5f).toType(DataType.FLOAT32, false);
     *   NDArray correct   = predicted.eq(y);
     *   return correct.mean().getFloat();
     * }</pre>
     *
     * <p>Acceptance threshold: returned accuracy must be >= 0.90 after training.
     *
     * @param model   trained DJL Model
     * @param manager NDManager for temporary sub-arrays
     * @param X       feature tensor, shape (nSamples, 2)
     * @param y       label tensor,   shape (nSamples, 1), values in {0.0, 1.0}
     * @return accuracy in [0, 1]
     */
    public static float evaluateAccuracy(Model model, NDManager manager, NDArray X, NDArray y) {
        // TODO: implement this
        throw new UnsupportedOperationException("TODO: implement this — see Javadoc above");
    }

    // ── Main ─────────────────────────────────────────────────────────────────

    public static void main(String[] args) throws Exception {
        System.out.println("=== 3.1.1 — Build Basic PyTorch MLP (DJL) ===\n");

        try (NDManager manager = NDManager.newBaseManager()) {

            // Implemented helper — generate synthetic moons data
            NDArray[] data = generateMoonsData(manager, 1000, 0.2f, 42L);
            NDArray X = data[0];  // shape (1000, 2)
            NDArray y = data[1];  // shape (1000, 1)
            System.out.printf("Dataset shape: X=%s  y=%s%n%n", X.getShape(), y.getShape());

            // Task 1 — build model block
            SequentialBlock block = buildMlp(/* hiddenDim= */ 64, /* outputDim= */ 1);
            System.out.println("Model block built successfully.");

            // Configure DJL trainer: Adam optimizer + sigmoid BCE loss + accuracy evaluator
            TrainingConfig config = new DefaultTrainingConfig(
                        Loss.sigmoidBinaryCrossEntropyLoss())
                    .optOptimizer(
                        Optimizer.adam()
                            .optLearningRateTracker(Tracker.fixed(0.001f))
                            .build())
                    .addEvaluator(new BinaryAccuracy())
                    .addTrainingListeners(TrainingListener.Defaults.logging());

            try (Model model = Model.newInstance("mlp-3.1.1")) {
                model.setBlock(block);

                try (Trainer trainer = model.newTrainer(config)) {
                    // Initialize weights — DJL defers allocation until first shape is known
                    trainer.initialize(new Shape(1, 2));
                    System.out.println("Trainer initialized.\n");

                    // Task 2 — full training loop
                    List<Float> trainLosses = trainModel(
                            trainer, X, y, /* epochs= */ 50, /* batchSize= */ 32);

                    System.out.printf("%nFinal Train Loss: %.4f%n", trainLosses.get(trainLosses.size() - 1));

                    // Task 3 — evaluate final accuracy
                    float accuracy = evaluateAccuracy(model, manager, X, y);
                    printMetric("Final Accuracy (full dataset)", accuracy);

                    if (accuracy >= 0.90f) {
                        System.out.println("\n✓ PASS — accuracy >= 0.90");
                    } else {
                        System.out.printf("%n✗ FAIL — accuracy %.4f < 0.90%n", accuracy);
                    }
                }
            }
        }
    }
}
